File: Leetcode/Python/ReverseWordsinString.py
__author__ = 'zhangruichang'
class Solution(object):
    def reverseWords(self, s):
        """
        :type s: a list of 1 length strings (List[str])
        :rtype: nothing
        """
        #if len(s) == 0:
        #    return '';
        s.reverse()
        start,n= 0, len(s)
        for i in range(len(s)):
            if s[i]==' ':
                s[start:i]=s[start:i][::-1]
                start = i+1
        s[start:n]=s[start:n][::-1]

S = Solution()
s = 'I am Zhang Ruichang'
s = [i for i in s]
print s
S.reverseWords(s)
print s

File: Leetcode/C++/ReverseWordsInString.cpp
#include<iostream>
#include <string>
using namespace std;


void rev(string& s, int st, int en){
    for(int i=st, j=en; i<j ;i++, j--){
        swap(s[i], s[j]);
    }
}

void reversestring(string& s){
    int n=s.size();
    rev(s, 0, n-1);
    for(int i=0, start=0; i<n;i++){
        if(s[i]==' '){
            rev(s, start, i-1);
            start=i+1;
        }
    }
}


int main(){
    string s="I am Zhang Ruichang";
    reversestring(s);
    cout<<s<<endl;
}